File: src/frontend/nfe/lib/app/data/model/nfe_det_especifico_combustivel_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/data/domain/domain_imports.dart';

class NfeDetEspecificoCombustivelModel {
	int? id;
	int? idNfeDetalhe;
	int? codigoAnp;
	String? descricaoAnp;
	double? percentualGlp;
	double? percentualGasNacional;
	double? percentualGasImportado;
	double? valorPartida;
	String? codif;
	double? quantidadeTempAmbiente;
	String? ufConsumo;
	double? cideBaseCalculo;
	double? cideAliquota;
	double? cideValor;
	int? encerranteBico;
	int? encerranteBomba;
	int? encerranteTanque;
	double? encerranteValorInicio;
	double? encerranteValorFim;

	NfeDetEspecificoCombustivelModel({
		this.id,
		this.idNfeDetalhe,
		this.codigoAnp,
		this.descricaoAnp,
		this.percentualGlp,
		this.percentualGasNacional,
		this.percentualGasImportado,
		this.valorPartida,
		this.codif,
		this.quantidadeTempAmbiente,
		this.ufConsumo,
		this.cideBaseCalculo,
		this.cideAliquota,
		this.cideValor,
		this.encerranteBico,
		this.encerranteBomba,
		this.encerranteTanque,
		this.encerranteValorInicio,
		this.encerranteValorFim,
	});

	static List<String> dbColumns = <String>[
		'id',
		'codigo_anp',
		'descricao_anp',
		'percentual_glp',
		'percentual_gas_nacional',
		'percentual_gas_importado',
		'valor_partida',
		'codif',
		'quantidade_temp_ambiente',
		'uf_consumo',
		'cide_base_calculo',
		'cide_aliquota',
		'cide_valor',
		'encerrante_bico',
		'encerrante_bomba',
		'encerrante_tanque',
		'encerrante_valor_inicio',
		'encerrante_valor_fim',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Codigo Anp',
		'Descricao Anp',
		'Percentual Glp',
		'Percentual Gas Nacional',
		'Percentual Gas Importado',
		'Valor Partida',
		'Codif',
		'Quantidade Temp Ambiente',
		'Uf Consumo',
		'Cide Base Calculo',
		'Cide Aliquota',
		'Cide Valor',
		'Encerrante Bico',
		'Encerrante Bomba',
		'Encerrante Tanque',
		'Encerrante Valor Inicio',
		'Encerrante Valor Fim',
	];

	NfeDetEspecificoCombustivelModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeDetalhe = jsonData['idNfeDetalhe'];
		codigoAnp = jsonData['codigoAnp'];
		descricaoAnp = jsonData['descricaoAnp'];
		percentualGlp = jsonData['percentualGlp']?.toDouble();
		percentualGasNacional = jsonData['percentualGasNacional']?.toDouble();
		percentualGasImportado = jsonData['percentualGasImportado']?.toDouble();
		valorPartida = jsonData['valorPartida']?.toDouble();
		codif = jsonData['codif'];
		quantidadeTempAmbiente = jsonData['quantidadeTempAmbiente']?.toDouble();
		ufConsumo = NfeDetEspecificoCombustivelDomain.getUfConsumo(jsonData['ufConsumo']);
		cideBaseCalculo = jsonData['cideBaseCalculo']?.toDouble();
		cideAliquota = jsonData['cideAliquota']?.toDouble();
		cideValor = jsonData['cideValor']?.toDouble();
		encerranteBico = jsonData['encerranteBico'];
		encerranteBomba = jsonData['encerranteBomba'];
		encerranteTanque = jsonData['encerranteTanque'];
		encerranteValorInicio = jsonData['encerranteValorInicio']?.toDouble();
		encerranteValorFim = jsonData['encerranteValorFim']?.toDouble();
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeDetalhe'] = idNfeDetalhe != 0 ? idNfeDetalhe : null;
		jsonData['codigoAnp'] = codigoAnp;
		jsonData['descricaoAnp'] = descricaoAnp;
		jsonData['percentualGlp'] = percentualGlp;
		jsonData['percentualGasNacional'] = percentualGasNacional;
		jsonData['percentualGasImportado'] = percentualGasImportado;
		jsonData['valorPartida'] = valorPartida;
		jsonData['codif'] = codif;
		jsonData['quantidadeTempAmbiente'] = quantidadeTempAmbiente;
		jsonData['ufConsumo'] = NfeDetEspecificoCombustivelDomain.setUfConsumo(ufConsumo);
		jsonData['cideBaseCalculo'] = cideBaseCalculo;
		jsonData['cideAliquota'] = cideAliquota;
		jsonData['cideValor'] = cideValor;
		jsonData['encerranteBico'] = encerranteBico;
		jsonData['encerranteBomba'] = encerranteBomba;
		jsonData['encerranteTanque'] = encerranteTanque;
		jsonData['encerranteValorInicio'] = encerranteValorInicio;
		jsonData['encerranteValorFim'] = encerranteValorFim;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeDetalhe = plutoRow.cells['idNfeDetalhe']?.value;
		codigoAnp = plutoRow.cells['codigoAnp']?.value;
		descricaoAnp = plutoRow.cells['descricaoAnp']?.value;
		percentualGlp = plutoRow.cells['percentualGlp']?.value?.toDouble();
		percentualGasNacional = plutoRow.cells['percentualGasNacional']?.value?.toDouble();
		percentualGasImportado = plutoRow.cells['percentualGasImportado']?.value?.toDouble();
		valorPartida = plutoRow.cells['valorPartida']?.value?.toDouble();
		codif = plutoRow.cells['codif']?.value;
		quantidadeTempAmbiente = plutoRow.cells['quantidadeTempAmbiente']?.value?.toDouble();
		ufConsumo = plutoRow.cells['ufConsumo']?.value != '' ? plutoRow.cells['ufConsumo']?.value : 'AC';
		cideBaseCalculo = plutoRow.cells['cideBaseCalculo']?.value?.toDouble();
		cideAliquota = plutoRow.cells['cideAliquota']?.value?.toDouble();
		cideValor = plutoRow.cells['cideValor']?.value?.toDouble();
		encerranteBico = plutoRow.cells['encerranteBico']?.value;
		encerranteBomba = plutoRow.cells['encerranteBomba']?.value;
		encerranteTanque = plutoRow.cells['encerranteTanque']?.value;
		encerranteValorInicio = plutoRow.cells['encerranteValorInicio']?.value?.toDouble();
		encerranteValorFim = plutoRow.cells['encerranteValorFim']?.value?.toDouble();
	}	

	NfeDetEspecificoCombustivelModel clone() {
		return NfeDetEspecificoCombustivelModel(
			id: id,
			idNfeDetalhe: idNfeDetalhe,
			codigoAnp: codigoAnp,
			descricaoAnp: descricaoAnp,
			percentualGlp: percentualGlp,
			percentualGasNacional: percentualGasNacional,
			percentualGasImportado: percentualGasImportado,
			valorPartida: valorPartida,
			codif: codif,
			quantidadeTempAmbiente: quantidadeTempAmbiente,
			ufConsumo: ufConsumo,
			cideBaseCalculo: cideBaseCalculo,
			cideAliquota: cideAliquota,
			cideValor: cideValor,
			encerranteBico: encerranteBico,
			encerranteBomba: encerranteBomba,
			encerranteTanque: encerranteTanque,
			encerranteValorInicio: encerranteValorInicio,
			encerranteValorFim: encerranteValorFim,
		);			
	}

	
}

File: backend/nfe/src/model/nfe_det_especifico_combustivel_model.py
from src import db


class NfeDetEspecificoCombustivelModel(db.Model):
    __tablename__ = 'nfe_det_especifico_combustivel'

    id = db.Column(db.Integer, primary_key=True)
    codigo_anp = db.Column(db.Integer)
    descricao_anp = db.Column(db.String(95))
    percentual_glp = db.Column(db.Float)
    percentual_gas_nacional = db.Column(db.Float)
    percentual_gas_importado = db.Column(db.Float)
    valor_partida = db.Column(db.Float)
    codif = db.Column(db.String(21))
    quantidade_temp_ambiente = db.Column(db.Float)
    uf_consumo = db.Column(db.String(2))
    cide_base_calculo = db.Column(db.Float)
    cide_aliquota = db.Column(db.Float)
    cide_valor = db.Column(db.Float)
    encerrante_bico = db.Column(db.Integer)
    encerrante_bomba = db.Column(db.Integer)
    encerrante_tanque = db.Column(db.Integer)
    encerrante_valor_inicio = db.Column(db.Float)
    encerrante_valor_fim = db.Column(db.Float)
    id_nfe_detalhe = db.Column(db.Integer, db.ForeignKey('nfe_detalhe.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_detalhe = data.get('idNfeDetalhe')
        self.codigo_anp = data.get('codigoAnp')
        self.descricao_anp = data.get('descricaoAnp')
        self.percentual_glp = data.get('percentualGlp')
        self.percentual_gas_nacional = data.get('percentualGasNacional')
        self.percentual_gas_importado = data.get('percentualGasImportado')
        self.valor_partida = data.get('valorPartida')
        self.codif = data.get('codif')
        self.quantidade_temp_ambiente = data.get('quantidadeTempAmbiente')
        self.uf_consumo = data.get('ufConsumo')
        self.cide_base_calculo = data.get('cideBaseCalculo')
        self.cide_aliquota = data.get('cideAliquota')
        self.cide_valor = data.get('cideValor')
        self.encerrante_bico = data.get('encerranteBico')
        self.encerrante_bomba = data.get('encerranteBomba')
        self.encerrante_tanque = data.get('encerranteTanque')
        self.encerrante_valor_inicio = data.get('encerranteValorInicio')
        self.encerrante_valor_fim = data.get('encerranteValorFim')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeDetalhe': self.id_nfe_detalhe,
            'codigoAnp': self.codigo_anp,
            'descricaoAnp': self.descricao_anp,
            'percentualGlp': self.percentual_glp,
            'percentualGasNacional': self.percentual_gas_nacional,
            'percentualGasImportado': self.percentual_gas_importado,
            'valorPartida': self.valor_partida,
            'codif': self.codif,
            'quantidadeTempAmbiente': self.quantidade_temp_ambiente,
            'ufConsumo': self.uf_consumo,
            'cideBaseCalculo': self.cide_base_calculo,
            'cideAliquota': self.cide_aliquota,
            'cideValor': self.cide_valor,
            'encerranteBico': self.encerrante_bico,
            'encerranteBomba': self.encerrante_bomba,
            'encerranteTanque': self.encerrante_tanque,
            'encerranteValorInicio': self.encerrante_valor_inicio,
            'encerranteValorFim': self.encerrante_valor_fim,
        }